apps/explorer/src/pages/checkpoints/CheckpointsTable.tsx

```diff
@@ -2,52 +2,73 @@
 // SPDX-License-Identifier: Apache-2.0
 import { useRpcClient } from '@mysten/core';
 import { useQuery } from '@tanstack/react-query';
-import { useMemo, useState } from 'react';
+import { useEffect, useMemo, useState } from 'react';
 
 import { TableFooter } from '~/components/Table/TableFooter';
 import { TxTableCol } from '~/components/transactions/TxCardUtils';
 import { TxTimeType } from '~/components/tx-time/TxTimeType';
 import { CheckpointLink } from '~/ui/InternalLink';
-import { usePaginationStack } from '~/ui/Pagination';
+import { useBoundedPaginationStack } from '~/ui/Pagination';
 import { PlaceholderTable } from '~/ui/PlaceholderTable';
 import { TableCard } from '~/ui/TableCard';
 import { Text } from '~/ui/Text';
 
 interface CheckpointsTableProps {
     initialLimit: number;
+    initialCursor?: string;
+    maxCursor?: string;
     disablePagination?: boolean;
     refetchInterval?: number;
 }
 
 export function CheckpointsTable({
     initialLimit,
+    initialCursor,
+    maxCursor,
     disablePagination,
     refetchInterval,
 }: CheckpointsTableProps) {
     const rpc = useRpcClient();
     const [limit, setLimit] = useState(initialLimit);
+    const [cursor, setCursor] = useState(initialCursor);
 
     const countQuery = useQuery(['checkpoints', 'count'], () =>
         rpc.getLatestCheckpointSequenceNumber()
     );
+    const pagination = useBoundedPaginationStack<string>(
+        initialCursor,
+        maxCursor
+    );
 
-    const pagination = usePaginationStack<string>();
+    const count = useMemo(() => {
+        if (maxCursor && initialCursor) return +initialCursor - +maxCursor;
+        return Number(countQuery.data ?? 0);
+    }, [countQuery.data, initialCursor, maxCursor]);
 
     const { data: checkpointsData } = useQuery(
-        ['checkpoints', { limit, cursor: pagination.cursor }],
+        ['checkpoints', { limit, cursor }],
         () =>
             rpc.getCheckpoints({
-                limit,
+                limit:
+                    cursor && maxCursor && +cursor - +limit < +maxCursor
```

Comment:
No idea what this is. Lots of type conversion here too. Almost certainly needs a comment or two